rest/route/middleware.go

```diff
@@ -423,6 +423,48 @@ func (m *podAuthMiddleware) ServeHTTP(rw http.ResponseWriter, r *http.Request, n
 	next(rw, r)
 }
 
+type alertmanagerMiddleware struct{}
+
+// NewAlertmanagerMiddleware returns a middleware that verifies the request
+// is coming from Evergreen's configured Alertmanager Kanopy webhook.
+func NewAlertmanagerMiddleware() gimlet.Middleware {
+	return &alertmanagerMiddleware{}
+}
+
+func (m *alertmanagerMiddleware) ServeHTTP(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	username, password, ok := r.BasicAuth()
+	if !ok {
+		gimlet.WriteResponse(rw, gimlet.MakeJSONErrorResponder(gimlet.ErrorResponse{
+			StatusCode: http.StatusUnauthorized,
+			Message:    "not authorized",
+		}))
+		return
+	}
+	u, err := user.FindOneById(username)
+	if err != nil {
+		gimlet.WriteResponse(rw, gimlet.MakeJSONErrorResponder(gimlet.ErrorResponse{
+			StatusCode: http.StatusInternalServerError,
+			Message:    fmt.Sprintf("finding user '%s'", username),
+		}))
+		return
+	}
+	if u == nil {
+		gimlet.WriteResponse(rw, gimlet.MakeJSONErrorResponder(gimlet.ErrorResponse{
+			StatusCode: http.StatusNotFound,
+			Message:    fmt.Sprintf("user '%s' not found", username),
+		}))
+		return
+	}
+	if u.APIKey != password {
```

Comment:
Should there be an explicit set of permissions to ping the alertmanager route? It seems kind of weird if we would permit any user to ping this route. Maybe we could just roll it with the superuser permissions?

Reply:
I think only the alertmanager service user should be able to ever hit this route, since if we wanted to toggle degraded mode we could just click the flag in the UI ourselves. I changed this middleware to only accept if the request is coming from "alertmanager" rather than wrap the route in superuser permissions, because we need custom middleware for alertmanager anyway due to its basic auth requirement.